### PR TITLE
Clarify UB in `get_unchecked(_mut)`

### PR DESCRIPTION
**Pull request by scottmcm**
Inspired by #116915, it was unclear to me what exactly "out-of-bounds index" means in `get_unchecked`.

One could [potentially](https://rust.godbolt.org/z/hxM764orW) interpret it that `get_unchecked` is just another way to write `offset`, but I think `get_unchecked(len)` is supposed to be UB even though `.offet(len)` is well-defined (as is `.get_unchecked(..len)`), so write that more directly in the docs.

**libs-api folks**: Can you confirm whether this is what you expect this to mean?  And is the situation any different for `<*const [T]>::get_unchecked`?

### CODE REVIEW

**Comment by the8472** [comment]
>  And is the situation any different for <*const [T]>::get_unchecked?

To add: One could interpret this as an `offset`, in which case one-past-the-end would be fine. But if we reserve the right to add `intrinsics::assume(idx < len)` here then it should also be UB.

File: library/core/src/slice/mod.rs [hunk 1]
```diff
@@ -640,6 +640,11 @@ impl<T> [T] {
     /// Calling this method with an out-of-bounds index is *[undefined behavior]*
     /// even if the resulting reference is not used.
     ///
+    /// You can think of this like `.get(index).unwrap_unchecked()`.  It's UB
+    /// to call `.get_unchecked(len)`, even if you immediately convert to a
+    /// pointer.  And it's UB to call `.get_unchecked(..len +1)` or
+    /// `.get_unchecked(..=len)` similar.
```
Comment (alex) [inline library/core/src/slice/mod.rs:646]: I think you're missing a word here.

**Comment by asquared31415** [comment]
"Out of bounds index" is almost certainly meant to mean exactly the same cases where you get the following panic with `[idx]`:
```
thread 'main' panicked at src/main.rs:2:3:
index out of bounds: the len is 1 but the index is 10
```
This same "out of bounds" phrasing is used all over docs, the Reference, and a bunch of other places, it would be surprising for it to mean anything else.

**Comment by asquared31415** [comment]
Right, that's library UB not language UB for the raw pointer version.  I forgot that that method exists!

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting: abusing `get_unchecked_mut` to get a pointer to the end of a slice is clearly UB since it involves creating an out-of-bounds reference.

@rfcbot fcp merge

File: library/core/src/slice/mod.rs [hunk 2]
```diff
@@ -640,6 +640,11 @@ impl<T> [T] {
     /// Calling this method with an out-of-bounds index is *[undefined behavior]*
     /// even if the resulting reference is not used.
     ///
+    /// You can think of this like `.get(index).unwrap_unchecked()`.  It's UB
+    /// to call `.get_unchecked(len)`, even if you immediately convert to a
+    /// pointer.  And it's UB to call `.get_unchecked(..len +1)`,
```
Comment (ChayimFriedman2) [inline library/core/src/slice/mod.rs:645]: It should probably be either `.get_unchecked(..len + 1)` or `.get_unchecked(..len+1)` (notice the spaces).

Comment (asquared31415) [reply library/core/src/slice/mod.rs:645]: rustfmt formats as `(..len + 1)`, that should be used for consistency.

**Review by cuviper — APPROVED** [review]
LGTM!

@bors r+ rollup

**Pull request merged**